source/generators/SingleParticle.cc

```diff
@@ -111,56 +108,51 @@ void SingleParticle::SetParticleDefinition(G4String particle_name)
 
 void SingleParticle::GeneratePrimaryVertex(G4Event* event)
 {
-  // Generate an initial position for the particle using the geometry
-  G4ThreeVector position = geom_->GenerateVertex(region_);
-
-  // Particle generated at start-of-event
-  G4double time = 0.;
-
-  // Create a new vertex
-  G4PrimaryVertex* vertex = new G4PrimaryVertex(position, time);
-
   // Generate uniform random energy in [E_min, E_max]
-  G4double kinetic_energy = nexus::RandomEnergy(energy_max_,energy_min_);
+  G4double kinetic_energy = nexus::RandomEnergy(energy_max_, energy_min_);
 
-  // Generate random direction by default
-  G4ThreeVector momentum_direction_ = G4RandomDirection();
-
-    // Calculate cartesian components of momentum
+  // Calculate cartesian components of momentum
   G4double mass   = particle_definition_->GetPDGMass();
   G4double energy = kinetic_energy + mass;
   G4double pmod = std::sqrt(energy*energy - mass*mass);
-  G4double px = pmod * momentum_direction_.x();
-  G4double py = pmod * momentum_direction_.y();
-  G4double pz = pmod * momentum_direction_.z();
-
-  // If user provides a momentum direction, this one is used
-  if (momentum_X_ != 0. || momentum_Y_ != 0. || momentum_Z_ != 0.) {
-    // Normalize if needed
-    G4double mom_mod = std::sqrt(momentum_X_ * momentum_X_ +
-				 momentum_Y_ * momentum_Y_ +
-				 momentum_Z_ * momentum_Z_);
-    px = pmod * momentum_X_/mom_mod;
-    py = pmod * momentum_Y_/mom_mod;
-    pz = pmod * momentum_Z_/mom_mod;
-  } else if (costheta_min_ != -1. || costheta_max_ != 1. || phi_min_ != 0. || phi_max_ !=2.*pi) {
-    G4ThreeVector p = nexus::Direction(costheta_min_, costheta_max_, phi_min_, phi_max_);
-    px = p.x() * pmod;
-    py = p.y() * pmod;
-    pz = p.z() * pmod;
+
+  momentum_.setX(momentum_X_);
+  momentum_.setY(momentum_Y_);
+  momentum_.setZ(momentum_Z_);
```

Comment:
This should not be here. You should 

+ *completely remove* the three members `momentum_X/Y/Z_` from the class, replacing them with a single `G4ThreeVector momentum_` member.

+ In the constructor, replace 

   ```c++
   msg_ -> DeclareProperty("...", momentum_X_ , "..."); # replace these with ...
   msg_ -> DeclareProperty("...", momentum_[0], "..."); # ... ones like these
   ```

Reply:
Done.